core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java

```diff
@@ -291,137 +407,171 @@ Path versionHintFile() {
-            maxVersion = currentVersion;
-          }
-        }
-
-        return maxVersion;
-      } catch (IOException io) {
-        LOG.warn("Error trying to recover version-hint.txt data for {}", versionHintFile, e);
+  @VisibleForTesting
+  int findVersionWithOutVersionHint(FileSystem fs) {
+    try {
+      if (!fs.exists(metadataRoot())) {
+        // Either the table has just been created, or it has been corrupted, but either way, we have
+        // to start at version 0.
+        LOG.warn("Metadata for table not found in directory [{}]", metadataRoot());
         return 0;
       }
+      // List the metadata directory to find the version files, and try to recover the max
+      // available version
+      FileStatus[] files =
+          fs.listStatus(metadataRoot(), name -> VERSION_PATTERN.matcher(name.getName()).matches());
+      int maxVersion = 0;
+      for (FileStatus file : files) {
+        int currentVersion = version(file.getPath().getName());
+        if (currentVersion > maxVersion && getMetadataFile(currentVersion) != null) {
+          maxVersion = currentVersion;
+        }
+      }
+      return maxVersion;
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @VisibleForTesting
+  int findVersion() {
+    Path versionHintFile = versionHintFile();
+    FileSystem fs = getFileSystem(versionHintFile, conf);
+    try {
+      return fs.exists(versionHintFile)
+          ? findVersionByUsingVersionHint(fs, versionHintFile)
+          : findVersionWithOutVersionHint(fs);
+    } catch (Exception e) {
+      // try one last time
+      return findVersionWithOutVersionHint(fs);
     }
   }
 
   /**
-   * Renames the source file to destination, using the provided file system. If the rename failed,
-   * an attempt will be made to delete the source file.
+   * Renames the source file to destination, using the provided file system.
    *
    * @param fs the filesystem used for the rename
    * @param src the source file
    * @param dst the destination file
+   * @return If it returns true, then the commit was successful.
    */
-  private void renameToFinal(FileSystem fs, Path src, Path dst, int nextVersion) {
-    try {
-      if (!lockManager.acquire(dst.toString(), src.toString())) {
-        throw new CommitFailedException(
-            "Failed to acquire lock on file: %s with owner: %s", dst, src);
+  @VisibleForTesting
+  boolean commitNewVersion(
+      FileSystem fs, Path src, Path dst, Integer nextVersion, boolean supportGlobalLocking)
+      throws IOException {
+    if (fs.exists(dst)) {
+      throw new CommitFailedException("Version %d already exists: %s", nextVersion, dst);
+    }
+    int maxVersion = supportGlobalLocking ? findVersion() : findVersionWithOutVersionHint(fs);
+    if (!nextVersionIsLatest(nextVersion, maxVersion)) {
+      if (!supportGlobalLocking) {
+        io().deleteFile(versionHintFile().toString());
```

Comment:
For cases where the global locking service is not used, we always try to remove the wrong versionHintFile before throwing an exception.